File: frontend/src/lib/api/sets.ts
import { api } from "./client";

export interface SetArtist {
  id: string;
  name: string;
  slug: string;
  role: string;
}

export interface SetSummary {
  id: string;
  slug: string;
  title: string;
  date: string | null;
  duration_seconds: number | null;
  set_type: string;
  status: string;
  artists: SetArtist[];
  tags: string[];
}

export interface SetDetail extends SetSummary {
  description: string | null;
  audio_stream_url: string;
  waveform_url: string | null;
  normalized_lufs: number | null;
  venue: { id: string; name: string; slug: string } | null;
  party: { id: string; name: string; slug: string } | null;
  embed_allowed: boolean;
}

export interface SetState {
  position_seconds: number;
  completed: boolean;
  playback_rate?: number;
}

export const listSets = () =>
  api<{ items: SetSummary[]; total: number }>("/api/sets");

export const getSet = (slug: string) => api<SetDetail>(`/api/sets/${slug}`);

export const patchSet = (slug: string, patch: Partial<SetDetail>) =>
  api<SetDetail>(`/api/sets/${slug}`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });

export const getSetState = (slug: string) =>
  api<SetState>(`/api/sets/${slug}/state`);

export const putSetState = (slug: string, state: SetState) =>
  api<void>(`/api/sets/${slug}/state`, {
    method: "PUT",
    body: JSON.stringify(state),
  });

File: packages/core/src/setvault_core/schemas/sets.py
from __future__ import annotations

from datetime import date as _date
from typing import Literal

from pydantic import BaseModel

from setvault_core.schemas.catalog import PartyOut, VenueOut

SetType = Literal["opener", "closer", "b2b", "headline", "warmup", "unknown"]


class SetArtistOut(BaseModel):
    id: str
    name: str
    slug: str
    role: str


class SetSummaryOut(BaseModel):
    id: str
    slug: str
    title: str
    date: _date | None
    duration_seconds: int | None
    set_type: SetType
    status: str
    artists: list[SetArtistOut]
    tags: list[str]


class SetDetailOut(SetSummaryOut):
    party: PartyOut | None
    venue: VenueOut | None
    description: str | None
    audio_stream_url: str
    waveform_url: str | None
    normalized_lufs: float | None
    embed_allowed: bool


class SetListOut(BaseModel):
    items: list[SetSummaryOut]
    total: int


class SetPatchIn(BaseModel):
    title: str | None = None
    date: _date | None = None
    set_type: SetType | None = None
    party_id: str | None = None
    venue_id: str | None = None
    description: str | None = None
    artist_ids: list[str] | None = None
    tag_names: list[str] | None = None
